packages/fuels-programs/src/calls/call_handler.rs

```diff
@@ -0,0 +1,490 @@
+        let tx = self.build_tx().await?;
+        let provider = self.account.try_provider()?;
+
+        let tx_id = provider.send_transaction(tx).await?;
+        self.cached_tx_id = Some(tx_id);
+
+        Ok(SubmitResponse::<A, Vec<ContractCall>, ()>::new(tx_id, self))
+    }
+
+    /// Call contract methods on the node, in a simulated manner, meaning the state of the
+    /// blockchain is *not* modified but simulated.
+    /// It is the same as the [call] method because the API is more user-friendly this way.
+    ///
+    /// [call]: Self::call
+    pub async fn simulate<T: Tokenizable + Debug>(&mut self) -> Result<CallResponse<T>> {
+        self.call_or_simulate(true).await
+    }
+
+    async fn call_or_simulate<T: Tokenizable + Debug>(
+        &mut self,
+        simulate: bool,
+    ) -> Result<CallResponse<T>> {
+        let tx = self.build_tx().await?;
+        let provider = self.account.try_provider()?;
+
+        self.cached_tx_id = Some(tx.id(provider.chain_id()));
+
+        let tx_status = if simulate {
+            provider.dry_run(tx).await?
+        } else {
+            provider.send_transaction_and_await_commit(tx).await?
+        };
+
+        let receipts = tx_status.take_receipts_checked(Some(&self.log_decoder))?;
+
+        self.get_response(receipts)
+    }
+
+    /// Simulates a call without needing to resolve the generic for the return type
+    async fn simulate_without_decode(&self) -> Result<()> {
+        let provider = self.account.try_provider()?;
+        let tx = self.build_tx().await?;
+
+        provider.dry_run(tx).await?.check(None)?;
+
+        Ok(())
+    }
+
+    /// Create a [`CallResponse`] from call receipts
+    pub fn get_response<T: Tokenizable + Debug>(
+        &self,
+        receipts: Vec<Receipt>,
+    ) -> Result<CallResponse<T>> {
+        let mut receipt_parser = ReceiptParser::new(&receipts, self.decoder_config);
+
+        let final_tokens = self
+            .call
+            .iter()
+            .map(|call| receipt_parser.parse_call(&call.contract_id, &call.output_param))
+            .collect::<Result<Vec<_>>>()?;
+
+        let tokens_as_tuple = Token::Tuple(final_tokens);
+        let response = CallResponse::<T>::new(
+            T::from_token(tokens_as_tuple)?,
+            receipts,
+            self.log_decoder.clone(),
+            self.cached_tx_id,
+        );
+
+        Ok(response)
+    }
+}
+
+#[async_trait::async_trait]
+impl<A> TxDependencyExtension for CallHandler<A, Vec<ContractCall>, ()>
+where
+    A: Account,
+{
+    async fn simulate(&mut self) -> Result<()> {
+        self.simulate_without_decode().await?;
+
+        Ok(())
+    }
+
+    fn append_variable_outputs(mut self, num: u64) -> Self {
+        self.call
+            .iter_mut()
+            .take(1)
+            .for_each(|call| call.append_variable_outputs(num));
+
+        self
+    }
+
+    fn append_contract(mut self, contract_id: Bech32ContractId) -> Self {
```

Comment:
@Br1ght0ne, @segfault-magnet. Now the `append_external_contract` trait method returns `Result<Self>` and in the case that no calls are present in the multi-call case it will return an error.